File: smallestFromLeaf.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val=val 
        self.left=left
        self.right=right


class Solution:

# Approach 1- DFS (Recursive) [using implict call-stack]
    def smallestFromLeaf(self, root): # TC O(n) // SC O(n)

        def dfs(root):
            nonlocal ans
            if not root: return 
            
            sb.append(chr(ord('a')+root.val))
            
            if not root.left and not root.right:
                ans=min(ans, "".join(reversed(sb)))
            
            dfs(root.left)
            dfs(root.right)
            sb.pop()
    
    
        sb=[]
        ans='~'
        dfs(root)
        return ans

root=TreeNode(0)
root.left=TreeNode(1)
root.right=TreeNode(2)
root.left.left=TreeNode(3)
root.left.right=TreeNode(4)
root.right.left=TreeNode(3)
root.right.right=TreeNode(4)
s=Solution()
print(s.smallestFromLeaf(root))

File: smallestFromLeaf.java
import java.util.*;
// Approach 1- DFS (Recursive) [using implict call-stack]
class Solution { // TC O(n) SC O(n)
    String ans="~";
    public String smallestFromLeaf(Node root) {
        dfs(root, new StringBuilder());
        return ans;
    }
    
    public void dfs(Node root, StringBuilder sb){
        if(root==null) return;
        sb.append((char) ('a'+root.val));
        if(root.left==null && root.right==null){
            sb.reverse();
            String s=sb.toString();
            sb.reverse();
            
            if(s.compareTo(ans)<0){
                ans=s;
            }
        }
        dfs(root.left, sb);
        dfs(root.right, sb);
        sb.deleteCharAt(sb.length()-1);
    }
}

class smallestFromLeaf{
    public static void main(String[] args) {
        BinaryTree bt= new BinaryTree();
        bt.root=new Node(0);
        bt.root.left=new Node(1);
        bt.root.right=new Node(2);
        bt.root.left.left=new Node(3);
        bt.root.left.right=new Node(4);
        bt.root.right.left=new Node(3);
        bt.root.right.right=new Node(4);
        Solution s= new Solution();
        System.out.println(s.smallestFromLeaf(bt.root));

    }
}